File: Sample Codes/Visual Basic .Net/DeviceProgramming/DeviceProgramming/DeviceProgramming.vb
Public Class DeviceProgramming
    Public retCode As Integer
    Public hReader As Integer
    Dim connActive As Boolean

    Private Sub GroupBox3_Enter(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles GroupBox3.Enter

    End Sub

    Private Sub RadioButton2_CheckedChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles rbOff1.CheckedChanged

    End Sub

    Private Sub DeviceProgramming_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        InitMenu()
    End Sub
    Private Sub InitMenu()
        Dim i As Integer

        mMsg.Text = ""

        DisplayMsg(0, 0, "Program Ready")

        For i = 1 To 10
            cbPort.Items.Add("COM" & i)
        Next i
        cbPort.SelectedIndex = 0

        cbBaudRate.Items.Add("9600")
        cbBaudRate.Items.Add("115200")

        cbBaudRate.SelectedIndex = 0

        bConn.Enabled = True

        cT1.Checked = True
        cT2.Checked = True

        rbOn1.Checked = True
        rbOn2.Checked = True
        rbOn3.Checked = True
        rbOn4.Checked = True


        EnableAll(False)

        connActive = False

    End Sub
    Private Sub EnableAll(ByVal opt As Boolean)

        bSetBaud.Enabled = opt
        bSetTimeout.Enabled = opt
        bSetLedBuzz.Enabled = opt
        bReset.Enabled = opt

        tbT1.Enabled = opt
        tbT2.Enabled = opt
        tbTimes.Enabled = opt

        cT1.Enabled = opt
        cT2.Enabled = opt
        cbUpdate1.Enabled = opt
        cbUpdate2.Enabled = opt
        cbBlink1.Enabled = opt
        cbBlink2.Enabled = opt

        rbOn1.Enabled = opt
        rbOn2.Enabled = opt
        rbOn3.Enabled = opt
        rbOn4.Enabled = opt
        rbOff1.Enabled = opt
        rbOff2.Enabled = opt
        rbOff3.Enabled = opt
        rbOff4.Enabled = opt


    End Sub
    Public Sub DisplayMsg(ByVal mType As Object, ByVal msgCode As Integer, ByVal PrintText As String)

        Select Case mType
            'Notification
            Case 0
                mMsg.SelectionColor = System.Drawing.ColorTranslator.FromOle(&H4000)
                'Error Message
            Case 1
                mMsg.SelectionColor = System.Drawing.Color.Red
                'input data
            Case 2
                mMsg.SelectionColor = System.Drawing.Color.Black
                PrintText = "< " & PrintText

                'output data
            Case 3
                mMsg.SelectionColor = System.Drawing.Color.Black
                PrintText = "> " & PrintText

        End Select

        mMsg.SelectedText = PrintText & vbCrLf
        mMsg.SelectionStart = Len(mMsg.Text)
        mMsg.SelectionColor = System.Drawing.Color.Black
        mMsg.Focus()



    End Sub

    Private Sub bConn_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles bConn.Click
        Dim a, tmpStr As String
        Dim fwVersion(0 To 262) As Byte
        Dim fwLen, i As Integer

        a = cbPort.Text
        fwLen = 256
        tmpStr = ""

        retCode = ACR122_Open(a, hReader)

        If retCode = 0 Then
            DisplayMsg(0, 0, "Connection to " & a & " success")
            connActive = True
            bConn.Enabled = False
            bReset.Enabled = True

            retCode = ACR122_GetFirmwareVersion(hReader, 0, fwVersion(0), fwLen)

            If retCode = 0 Then
                DisplayMsg(0, 0, "Get Firmware Success")

                For i = 0 To fwLen - 1
                    tmpStr = tmpStr & Chr(fwVersion(i))
                Next i

                DisplayMsg(0, 0, "Firmware: " & tmpStr)
                DisplayMsg(0, 0, "")
            Else
                DisplayMsg(1, 0, "Get Firmware Failed")
            End If

            EnableAll(True)

        Else
            DisplayMsg(1, 0, "Connection to " & a & " failed")
        End If
    End Sub

    Private Sub bQuit_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles bQuit.Click
        If connActive = True Then
            ACR122_Close(hReader)
        End If

        Me.Close()
    End Sub

    Private Sub bReset_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles bReset.Click
        If connActive = True Then
            ACR122_Close(hReader)
            InitMenu()
            connActive = False
        End If
    End Sub

    Private Sub bClear_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles bClear.Click
        mMsg.Text = ""
    End Sub

    Private Sub bSetBaud_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles bSetBaud.Click

        retCode = ACR122_SetBaudRate(hReader, Int(cbBaudRate.Text))

        If retCode = 0 Then
            Call DisplayMsg(0, 0, "Set Baud Rate " & cbBaudRate.Text & " Success.")
        Else
            Call DisplayMsg(1, 0, "Set Baud Rate " & cbBaudRate.Text & " Failed")
        End If

    End Sub

    Private Sub tbT1_KeyPress(ByVal sender As System.Object, ByVal e As System.Windows.Forms.KeyPressEventArgs) Handles tbT1.KeyPress, tbT2.KeyPress, tbTimes.KeyPress
        If InStr("0123456789", e.KeyChar) > 0 Or e.KeyChar = Microsoft.VisualBasic.vbBack Then

        Else
            e.KeyChar = ""

        End If
    End Sub

    Private Sub bSetLedBuzz_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles bSetLedBuzz.Click
        Dim ledCtrl(1) As ACR122_LED_CONTROL
        Dim buzzMode, t1 As Object
        Dim t2 As Integer
        Dim num As Short

        If tbT1.Text = "" Then
            tbT1.Focus()
            Exit Sub
        End If

        If tbT2.Text = "" Then
            tbT2.Focus()
            Exit Sub
        End If

        If tbTimes.Text = "" Then
            tbTimes.Focus()
            Exit Sub
        End If

        If CShort(tbT1.Text) > 25500 Then
            tbT1.Text = "25500"
            tbT1.Focus()
            Exit Sub
        End If

        If CShort(tbT2.Text) > 25500 Then
            tbT2.Text = "25500"
            tbT2.Focus()
            Exit Sub
        End If

        If CShort(tbTimes.Text) > 255 Then
            tbTimes.Text = "255"
            tbTimes.Focus()
            Exit Sub
        End If

        num = CShort(tbTimes.Text)
        t1 = CInt(tbT1.Text)
        t2 = CInt(tbT2.Text)

        'L0 Final State
        If rbOn1.Checked = True Then
            ledCtrl(0).finalState = ACR122_LED_STATE_ON
        Else
            ledCtrl(0).finalState = ACR122_LED_STATE_OFF
        End If

        'L0 Initial Blinking State
        If rbOn2.Checked = True Then
            ledCtrl(0).initialBlinkingState = ACR122_LED_STATE_ON
        Else
            ledCtrl(0).initialBlinkingState = ACR122_LED_STATE_OFF
        End If

        'L1 Final State
        If rbOn3.Checked = True Then
            ledCtrl(1).finalState = ACR122_LED_STATE_ON
        Else
            ledCtrl(1).finalState = ACR122_LED_STATE_OFF
        End If

        'L1 Initial Blinking State
        If rbOn4.Checked = True Then
            ledCtrl(1).initialBlinkingState = ACR122_LED_STATE_ON
        Else
            ledCtrl(1).initialBlinkingState = ACR122_LED_STATE_OFF
        End If

        'L0 Enable Blink
        If cbBlink1.Checked = True Then
            ledCtrl(0).blinkEnabled = True
        Else
            ledCtrl(0).blinkEnabled = False
        End If

        'L1 Enable Blink
        If cbBlink2.Checked = True Then
            ledCtrl(1).blinkEnabled = True
        Else
            ledCtrl(1).blinkEnabled = False
        End If

        'L0 Enable Update
        If cbUpdate1.Checked = True Then
            ledCtrl(0).updateEnabled = True
        Else
            ledCtrl(0).updateEnabled = False
        End If

        'L1 Enable Update
        If cbUpdate2.Checked = True Then
            ledCtrl(1).updateEnabled = True
        Else
            ledCtrl(1).updateEnabled = False
        End If


        If cT1.Checked = True Then
            buzzMode = ACR122_BUZZER_MODE_T1
        End If

        If cT2.Checked = True Then
            buzzMode = ACR122_BUZZER_MODE_T2
        End If

        If cT1.Checked = True And cT2.Checked = True Then
            buzzMode = ACR122_BUZZER_MODE_T1 Or ACR122_BUZZER_MODE_T2
        End If

        If cT1.Checked = False And cT2.Checked = False Then
            buzzMode = ACR122_BUZZER_MODE_OFF
        End If

        retCode = ACR122_SetLedStatesWithBeep(hReader, ledCtrl(0), 2, t1, t2, num, buzzMode)

        If retCode = 0 Then
            DisplayMsg(0, 0, "Set LED States with Beep success")
        Else
            DisplayMsg(1, 0, "Set LED States with Beep failed")
        End If



    End Sub

    Private Sub Label2_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Label2.Click

    End Sub

    Private Sub GroupBox1_Enter(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles GroupBox1.Enter

    End Sub

    Private Sub bSetTimeout_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles bSetTimeout.Click
        Dialog.Show()

    End Sub
End Class

File: Sample Codes/Delphi/Device Programming/DeviceProgramming.pas
unit DeviceProgramming;

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms,
  ACR122s, Dialog, Dialogs, StdCtrls, ComCtrls;

type
    TDeviceProgram = class(TForm)
    Label1: TLabel;
    Label2: TLabel;
    cbPort: TComboBox;
    cbBaudRate: TComboBox;
    bConnect: TButton;
    bSetBaud: TButton;
    bSetTimeout: TButton;
    GroupBox1: TGroupBox;
    GroupBox2: TGroupBox;
    Label3: TLabel;
    Label4: TLabel;
    Label5: TLabel;
    Label6: TLabel;
    tbT1: TEdit;
    tbT2: TEdit;
    tbTimes: TEdit;
    cT1: TCheckBox;
    cT2: TCheckBox;
    GroupBox3: TGroupBox;
    cEnableUp1: TCheckBox;
    cEnableBlink1: TCheckBox;
    bSetLedBuzz: TButton;
    bClear: TButton;
    bReset: TButton;
    bQuit: TButton;
    mMsg: TRichEdit;
    GroupBox6: TGroupBox;
    rbOn1: TRadioButton;
    rbOff1: TRadioButton;
    Label7: TLabel;
    GroupBox5: TGroupBox;
    Label8: TLabel;
    rbOn2: TRadioButton;
    rbOff2: TRadioButton;
    GroupBox4: TGroupBox;
    cEnableUp2: TCheckBox;
    cEnableBlink2: TCheckBox;
    GroupBox7: TGroupBox;
    Label9: TLabel;
    rbOn3: TRadioButton;
    rbOff3: TRadioButton;
    GroupBox8: TGroupBox;
    Label10: TLabel;
    rbOn4: TRadioButton;
    rbOff4: TRadioButton;
    procedure FormActivate(Sender: TObject);
    procedure bConnectClick(Sender: TObject);
    procedure bSetBaudClick(Sender: TObject);
    procedure bQuitClick(Sender: TObject);
    procedure bResetClick(Sender: TObject);
    procedure bClearClick(Sender: TObject);
    procedure bSetLedBuzzClick(Sender: TObject);
    procedure bSetTimeoutClick(Sender: TObject);
    procedure tbT1KeyPress(Sender: TObject; var Key: Char);
  private
    { Private declarations }
  public

  end;

var
  DeviceProgram: TDeviceProgram;
  hReader : SCARDHANDLE;
  connActive :BOOL;
  PrintText : String;
  SendBuff : array [0..256] of Byte;
  RecvBuff : array [0..256] of Byte;
  retCode  : DWORD;
  SendLen : smallint;
  RecvLen : DWORD;
  tmpBaud  : DWORD;

procedure InitMenu();
procedure displayOut(errType: Integer; retVal: Integer; PrintText: String);
procedure EnableAll(Option : BOOL);
implementation

{$R *.dfm}

procedure TDeviceProgram.bClearClick(Sender: TObject);
begin
     DeviceProgram.mMsg.Clear;
end;

procedure TDeviceProgram.bConnectClick(Sender: TObject);
var
FWLEN : DWORD;
tempstr : array [0..256] of char;
begin
    PrintText := DeviceProgram.cbPort.Text;

    retCode := ACR122_OpenA( PrintText, @hReader);
    if retCode = 0  then
    begin
      ConnActive := True;
      DeviceProgram.bConnect.Enabled := False;
      EnableAll(True);

      PrintText := 'Connection to ' + PrintText + ' success';
      displayOut( 0, 0, PrintText);

      PrintText := '';
      
      retCode := ACR122_GetFirmwareVersionA(hReader, 0, tempstr, @FWLEN);
      if retCode = 0 then
      begin
        PrintText := 'Firmware Version: ' + tempstr;
        displayOut(5, 0, PrintText);
      end
      else
      begin
        displayOut( 1, 0, 'Get Firmware Version failed');
      end;
    end
    else
    begin
      PrintText := 'Connection to ' + PrintText + ' failed';
      displayOut( 1, 0, PrintText);
    end;
end;

procedure TDeviceProgram.bQuitClick(Sender: TObject);
begin
    if ConnActive = true then
     begin
        retCode := ACR122_Close(hReader);
     end;
    Application.Terminate;
end;

procedure TDeviceProgram.bResetClick(Sender: TObject);
begin
     if ConnActive = true then
     begin
        retCode := ACR122_Close(hReader);
        if retCode = 0 then
        begin
             InitMenu();
        end;

     end;
end;

procedure TDeviceProgram.bSetBaudClick(Sender: TObject);
begin

    tmpBaud := StrToInt(cbBaudRate.Text);

    retCode := ACR122_SetBaudRate(hReader, tmpBaud);

    if retCode = 0 then
    begin
        DisplayOut( 0, 0, 'Set baud rate success');
    end
    else
        DisplayOut( 0, 0, 'Set baud rate failed');
    begin

    end;
end;

procedure TDeviceProgram.bSetLedBuzzClick(Sender: TObject);
var
ledCtrl : array [0 .. 2 ] of ACR122_LED_CONTROL;
t1 : Integer;
t2 : Integer;
num : Integer;
buzzmode : DWORD;
begin

   // L0 Final State
   if rbOn1.Checked = True then begin
      ledCtrl[0].finalState := ACR122_LED_STATE_ON;
   end
   else begin
      ledCtrl[0].finalState := ACR122_LED_STATE_OFF;
   end;

   // L0 Initial Blinking State
   if rbOn2.Checked = True then begin
      ledCtrl[0].initialBlinkingState := ACR122_LED_STATE_ON;
   end
   else begin
      ledCtrl[0].initialBlinkingState := ACR122_LED_STATE_OFF;
   end;

   // L1 Final State
   if rbOn3.Checked = True then begin
      ledCtrl[1].finalState := ACR122_LED_STATE_ON;
   end
   else begin
      ledCtrl[1].finalState := ACR122_LED_STATE_OFF;
   end;

   // L1 Initial Blinking State
   if rbOn4.Checked = True then begin
      ledCtrl[1].initialBlinkingState := ACR122_LED_STATE_ON;
   end
   else begin
      ledCtrl[1].initialBlinkingState := ACR122_LED_STATE_OFF;
   end;

   // L0 Enable Blink
   if cEnableBlink1.Checked = True then begin
      ledCtrl[0].blinkEnabled := True;
   end
   else begin
      ledCtrl[0].blinkEnabled := False;
   end;

   // L1 Enable Blink
   if cEnableBlink2.Checked = True then begin
      ledCtrl[1].blinkEnabled := True;
   end
   else begin
      ledCtrl[1].blinkEnabled := False;
   end;

   // L0 Enable Update
   if cEnableUp1.Checked = True then begin
      ledCtrl[0].updateEnabled := True;
   end
   else begin
      ledCtrl[0].updateEnabled := False;
   end;

   // L1 Enable Update
   if cEnableUp2.Checked = True then begin
      ledCtrl[1].updateEnabled := True;
   end
   else begin
      ledCtrl[1].updateEnabled := False;
   end;

   if tbT1.Text = '' then begin
      tbT1.SetFocus;
      Exit;
   end;

   if tbT2.Text = '' then begin
      tbT2.SetFocus;
      Exit;
   end;

   if tbTimes.Text = '' then begin
      tbTimes.SetFocus;
      Exit;
   end;

   t1 := StrToInt(tbT1.Text);
   if t1 > 25500 then begin
      tbT1.Text := '25500';
      tbT1.SetFocus;
      Exit;
   end;

   t2 := StrToInt(tbT2.Text);
   if t2 > 25500 then begin
      tbT2.Text := '25500';
      tbT2.SetFocus;
      Exit;
   end;

   num := StrToInt(tbTimes.Text);
   if num > 255 then begin
      tbTimes.Text := '255';
      tbTimes.SetFocus;
      Exit;
   end;

   if cT1.Checked = True then begin
     buzzmode := ACR122_BUZZER_MODE_ON_T1;
   end;

   if cT2.Checked = True then begin
     buzzmode := ACR122_BUZZER_MODE_ON_T2;
   end;

   if (cT1.Checked = True) and (cT2.Checked = True) then begin
     buzzmode := ACR122_BUZZER_MODE_ON_T1 or ACR122_BUZZER_MODE_ON_T2;
   end;

   if (cT1.Checked = False) and (cT2.Checked = False) then begin
     buzzmode := ACR122_BUZZER_MODE_OFF;
   end;

   retCode := ACR122_SetLedStatesWithBeep(hReader, @ledCtrl, 2 , t1, t2, num, buzzmode);

   if retCode = 0 then begin
      DisplayOut( 0, 0, 'Set LED States with Beep success');
   end
   else begin
      DisplayOut( 1, 0, 'Set LED States with Beep failed');
   end;


end;

procedure TDeviceProgram.bSetTimeoutClick(Sender: TObject);
begin
   FDialog.Show;
end;

procedure TDeviceProgram.FormActivate(Sender: TObject);
begin
   InitMenu();
end;

procedure TDeviceProgram.tbT1KeyPress(Sender: TObject; var Key: Char);
begin
  if Key <> chr($08) then
    if Not (Key in ['0'..'9'])then
      Key := Chr($00);
end;

procedure InitMenu();
var
indx : Integer;
begin

  connActive := False;

  DeviceProgram.mMsg.Clear;

  DisplayOut(0, 0, 'Program ready');

  DeviceProgram.cbPort.Clear;

  for indx := 1 to 10 do
  begin
    PrintText := 'COM' + IntToStr(indx);
    DeviceProgram.cbPort.AddItem( PrintText, TObject.NewInstance);
  end;

  DeviceProgram.cbPort.ItemIndex := 0;

  DeviceProgram.cbBaudRate.AddItem('9600', TObject.NewInstance);
  DeviceProgram.cbBaudRate.AddItem('115200', TObject.NewInstance);

  DeviceProgram.cbBaudRate.ItemIndex := 0;

  DeviceProgram.tbT1.Text := '';
  DeviceProgram.tbT2.Text := '';
  DeviceProgram.tbTimes.Text := '';

  DeviceProgram.cT1.Checked := True;
  DeviceProgram.cT2.Checked := True;

  DeviceProgram.rbON1.Checked := True;
  DeviceProgram.rbON2.Checked := True;
  DeviceProgram.rbON3.Checked := True;
  DeviceProgram.rbON4.Checked := True;

  DeviceProgram.bConnect.Enabled := True;

  EnableAll(False);

end;

procedure EnableAll(Option : BOOL);
begin

  DeviceProgram.bReset.Enabled := Option;
  DeviceProgram.bSetBaud.Enabled := Option;
  DeviceProgram.bSetTimeout.Enabled := Option;
  DeviceProgram.bSetLedBuzz.Enabled := Option;
  DeviceProgram.cT1.Enabled := Option;
  DeviceProgram.cT2.Enabled := Option;
  DeviceProgram.tbT1.Enabled := Option;
  DeviceProgram.tbT2.Enabled := Option;
  DeviceProgram.tbTimes.Enabled := Option;
  DeviceProgram.rbON1.Enabled := Option;
  DeviceProgram.rbON2.Enabled := Option;
  DeviceProgram.rbON3.Enabled := Option;
  DeviceProgram.rbON4.Enabled := Option;
  DeviceProgram.rbOFF1.Enabled := Option;
  DeviceProgram.rbOFF2.Enabled := Option;
  DeviceProgram.rbOFF3.Enabled := Option;
  DeviceProgram.rbOFF4.Enabled := Option;
  DeviceProgram.cEnableUp1.Enabled := Option;
  DeviceProgram.cEnableUp2.Enabled := Option;
  DeviceProgram.cEnableBlink1.Enabled := Option;
  DeviceProgram.cEnableBlink2.Enabled := Option;

end;

procedure displayOut(errType: Integer; retVal: Integer; PrintText: String);
begin

  case errType of
    0: DeviceProgram.mMsg.SelAttributes.Color := clTeal;      // Notifications
    1: begin                                                // Error Messages
         DeviceProgram.mMsg.SelAttributes.Color := clRed;
         //PrintText := GetScardErrMsg(retVal);
       end;
    2: begin
         DeviceProgram.mMsg.SelAttributes.Color := clBlack;
         PrintText := '< ' + PrintText;                      // Input data
       end;
    3: begin
         DeviceProgram.mMsg.SelAttributes.Color := clBlack;
         PrintText := '> ' + PrintText;                      // Output data
       end;
    4: DeviceProgram.mMsg.SelAttributes.Color := clRed;        // For ACOS1 error
    5: DeviceProgram.mMsg.SelAttributes.Color := clBlack;      // Normal Notification
  end;
  DeviceProgram.mMsg.Lines.Add(PrintText);
  DeviceProgram.mMsg.SelAttributes.Color := clBlack;
  DeviceProgram.mMsg.SetFocus;

end;

end.
